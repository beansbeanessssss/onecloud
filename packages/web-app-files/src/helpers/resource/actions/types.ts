export enum TransferType {
  COPY,
  MOVE,
  DUPLICATE
}
