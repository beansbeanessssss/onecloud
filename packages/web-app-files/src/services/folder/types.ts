export interface FolderLoaderOptions {
  shareIndicators?: boolean
}
