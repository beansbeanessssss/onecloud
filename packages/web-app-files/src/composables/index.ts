export * from './actions'
export * from './resourcesViewDefaults'
export * from './shares'
