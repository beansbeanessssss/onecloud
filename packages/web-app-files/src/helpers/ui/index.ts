export * as fileList from './filesList'
export * as resourceTiles from './resourceTiles'
