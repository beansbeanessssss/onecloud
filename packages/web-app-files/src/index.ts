import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import { defineWebApplication, quickActions } from '@ownclouders/web-pkg'
import store from './store'
import { extensions } from './extensions'
import fileSideBars from './fileSideBars'
import { buildRoutes } from '@ownclouders/web-pkg'
import { AppNavigationItem } from '@ownclouders/web-pkg'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import {
  SpaceResource,
  isPersonalSpaceResource,
  isShareSpaceResource
} from '@ownclouders/web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  color: 'var(--oc-color-swatch-primary-muted)',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
export const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      isActive: () => {
        return (
          !context.$store.getters['runtime/spaces/currentSpace'] ||
          context.$store.getters['runtime/spaces/currentSpace']?.isOwner(
            context.$store.getters.user
          )
        )
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find(
          (drive) => isPersonalSpaceResource(drive) && drive.isOwner(context.$store.getters.user)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      isActive: () => {
        const space = context.$store.getters['runtime/spaces/currentSpace'] as SpaceResource
        // last check is when fullShareOwnerPaths is enabled
        return !space || isShareSpaceResource(space) || !space?.isOwner(context.$store.getters.user)
      },
      activeFor: [
        { path: `/${appInfo.id}/spaces/share` },
        { path: `/${appInfo.id}/spaces/personal` }
      ],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0' && capabilities.files?.undelete
      }
    }
  ]
}

export default defineWebApplication({
  setup(args) {
    return {
      appInfo,
      store,
      routes: buildRoutes({
        App,
        Favorites,
        FilesDrop,
        SearchResults,
        Shares: {
          SharedViaLink,
          SharedWithMe,
          SharedWithOthers
        },
        Spaces: {
          DriveResolver: SpaceDriveResolver,
          Projects: SpaceProjects
        },
        Trash: {
          Overview: TrashOverview
        }
      }),
      navItems,
      quickActions,
      translations,
      extensions: extensions(args)
    }
  }
})
