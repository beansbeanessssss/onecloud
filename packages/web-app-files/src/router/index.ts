import {
  buildRoutes as buildSpacesRoutes,
  isLocationSpacesActive,
  createLocationSpaces
} from './spaces'
import {
  buildRoutes as buildSharesRoutes,
  isLocationSharesActive,
  createLocationShares
} from './shares'
import {
  buildRoutes as buildOperationsRoutes,
  createLocationOperations,
  isLocationOperationsActive
} from './operations'
import {
  buildRoutes as buildCommonRoutes,
  isLocationCommonActive,
  createLocationCommon
} from './common'
import { isRoutePublic, isLocationActive } from './utils'
import { buildRoutes as buildDeprecatedRoutes } from './deprecated'
import { RouteComponents } from './router'
import { RouteConfig } from 'vue-router'

const ROOT_ROUTE = {
  path: '/',
  redirect: (to) => createLocationSpaces(to)
}

const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  ROOT_ROUTE,
  ...buildSharesRoutes(components),
  ...buildSpacesRoutes(components),
  ...buildOperationsRoutes(components),
  ...buildCommonRoutes(components),
  ...buildDeprecatedRoutes()
]

// const isLocationOf = isLocationActiveCreator<locationTypes>(createLocationFor)

export {
  createLocationCommon,
  createLocationShares,
  createLocationSpaces,
  createLocationOperations,
  isLocationOperationsActive,
  isLocationCommonActive,
  isLocationSharesActive,
  isLocationSpacesActive,
  isLocationActive,
  isRoutePublic,
  buildRoutes
}
