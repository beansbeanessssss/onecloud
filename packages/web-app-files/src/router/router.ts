import Vue, { ComponentOptions } from 'vue'

/**
 * we need to inject the vue files into the route builders,
 * this is because we also import the provided helpers from other js|ts files
 * like mixins, rollup seems to have a problem to import files which contain vue file imports
 * into js files which then again get imported by other vue files...
 */
export interface RouteComponents {
  App: ComponentOptions<typeof Vue>
  Favorites: ComponentOptions<typeof Vue>
  FilesDrop: ComponentOptions<typeof Vue>
  PrivateLink: ComponentOptions<typeof Vue>
  SearchResults: ComponentOptions<typeof Vue>
  PublicLink: ComponentOptions<typeof Vue>
  Shares: {
    SharedWithMe: ComponentOptions<typeof Vue>
    SharedWithOthers: ComponentOptions<typeof Vue>
    SharedViaLink: ComponentOptions<typeof Vue>
  }
  Spaces: {
    DriveResolver: ComponentOptions<typeof Vue>
    Projects: ComponentOptions<typeof Vue>
  }
}
