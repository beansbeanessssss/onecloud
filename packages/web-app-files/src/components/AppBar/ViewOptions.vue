<template>
  <div class="oc-flex oc-flex-middle">
    <div
      v-if="viewModes.length"
      class="viewmode-switch-buttons oc-button-group oc-visible@s oc-mr-s"
    >
      <oc-button
        v-for="viewMode in viewModes"
        :key="viewMode.name"
        v-oc-tooltip="$gettext(viewMode.label)"
        :class="viewMode.name"
        :appearance="viewModeCurrent === viewMode.name ? 'filled' : 'outline'"
        :aria-label="$gettext(viewMode.label)"
        @click="setViewMode(viewMode)"
      >
        <oc-icon :name="viewMode.icon.name" :fill-type="viewMode.icon.fillType" size="small" />
      </oc-button>
    </div>
    <oc-button
      id="files-view-options-btn"
      key="files-view-options-btn"
      v-oc-tooltip="viewOptionsButtonLabel"
      data-testid="files-view-options-btn"
      :aria-label="viewOptionsButtonLabel"
      appearance="raw"
      class="oc-my-s oc-p-xs"
    >
      <oc-icon name="settings-3" fill-type="line" />
    </oc-button>
    <oc-drop
      drop-id="files-view-options-drop"
      toggle="#files-view-options-btn"
      mode="click"
      class="oc-width-auto"
      padding-size="medium"
    >
      <oc-list>
        <li class="files-view-options-list-item oc-mb-m">
          <oc-switch
            v-model:checked="hiddenFilesShownModel"
            data-testid="files-switch-hidden-files"
            :label="$gettext('Show hidden files')"
            @update:checked="updateHiddenFilesShownModel"
          />
        </li>
        <li class="files-view-options-list-item oc-my-m">
          <oc-switch
            v-model:checked="fileExtensionsShownModel"
            data-testid="files-switch-files-extensions-files"
            :label="$gettext('Show file extensions')"
            @update:checked="updateFileExtensionsShownModel"
          />
        </li>
        <li class="files-view-options-list-item oc-mt-m">
          <oc-page-size
            v-if="!queryParamsLoading"
            :selected="itemsPerPage"
            data-testid="files-pagination-size"
            :label="$gettext('Items per page')"
            :options="[100, 500]"
            class="files-pagination-size"
            @change="itemsPerPage = $event"
          />
        </li>
      </oc-list>
    </oc-drop>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType, ref, watch } from 'vue'
import { mapMutations, mapState } from 'vuex'
import { useRouteQueryPersisted } from 'web-pkg/src/composables'
import { ViewMode } from 'web-pkg/src/ui/types'
import { PaginationConstants, ViewModeConstants } from '../../composables'

export default defineComponent({
  props: {
    viewModes: {
      type: Array as PropType<ViewMode[]>,
      default: () => []
    }
  },
  setup() {
    const queryParamsLoading = ref(false)
    const perPageQuery = useRouteQueryPersisted({
      name: PaginationConstants.perPageQueryName,
      defaultValue: PaginationConstants.perPageDefault
    })
    const viewModeQuery = useRouteQueryPersisted({
      name: ViewModeConstants.queryName,
      defaultValue: ViewModeConstants.defaultModeName
    })
    watch(
      [perPageQuery, viewModeQuery],
      (params) => {
        queryParamsLoading.value = params.some((p) => !p)
      },
      { immediate: true, deep: true }
    )

    return {
      ViewModeConstants,
      viewModeCurrent: viewModeQuery,
      itemsPerPage: perPageQuery,
      queryParamsLoading
    }
  },
  computed: {
    ...mapState('Files', ['areHiddenFilesShown', 'areFileExtensionsShown']),

    viewOptionsButtonLabel() {
      return this.$gettext('Display customization options of the files list')
    },

    hiddenFilesShownModel: {
      get() {
        return this.areHiddenFilesShown
      },

      set(value) {
        this.SET_HIDDEN_FILES_VISIBILITY(value)
      }
    },
    fileExtensionsShownModel: {
      get() {
        return this.areFileExtensionsShown
      },

      set(value) {
        this.SET_FILE_EXTENSIONS_VISIBILITY(value)
      }
    }
  },
  methods: {
    ...mapMutations('Files', ['SET_HIDDEN_FILES_VISIBILITY', 'SET_FILE_EXTENSIONS_VISIBILITY']),
    setViewMode(mode) {
      this.viewModeCurrent = mode.name
    },
    updateHiddenFilesShownModel(event) {
      this.hiddenFilesShownModel = event
    },
    updateFileExtensionsShownModel(event) {
      this.fileExtensionsShownModel = event
    }
  }
})
</script>

<style lang="scss" scoped>
.viewmode-switch-buttons {
  flex-flow: initial;
}

#files-view-options-btn {
  vertical-align: middle;
  border: 3px solid transparent;
  &:hover {
    background-color: var(--oc-color-background-hover);
    border-radius: 3px;
  }
}
.files-view-options-list-item {
  & > * {
    display: flex;
    justify-content: space-between;
  }

  & + & {
    margin-top: var(--oc-space-small);
  }
}
</style>
