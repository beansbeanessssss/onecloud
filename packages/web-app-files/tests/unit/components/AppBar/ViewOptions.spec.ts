import ViewOptions from 'web-app-files/src/components/AppBar/ViewOptions.vue'
import { useRouteQueryPersisted } from 'web-pkg/src/composables/router'
import { ref } from 'vue'
import {
  createStore,
  defaultPlugins,
  defaultStoreMockOptions,
  defaultComponentMocks,
  mount
} from 'web-test-helpers'
import { ViewModeConstants } from 'web-app-files/src/composables'
import { ViewMode } from 'web-pkg/src/ui/types'

jest.mock('web-pkg/src/composables/router')
const selectors = {
  pageSizeSelect: '.oc-page-size',
  hiddenFilesSwitch: '[data-testid="files-switch-hidden-files"]',
  fileExtensionsSwitch: '[data-testid="files-switch-files-extensions-files"]'
}

describe('ViewOptions component', () => {
  it('sets the correct initial files page limit', () => {
    const perPage = '100'
    const { wrapper } = getWrapper({ perPage })
    expect(wrapper.findComponent<any>(selectors.pageSizeSelect).props().selected).toBe(perPage)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('toggles the setting to show/hide hidden files', () => {
    const { wrapper, storeOptions } = getWrapper()
    ;(wrapper.findComponent<any>(selectors.hiddenFilesSwitch).vm as any).$emit('change', false)
    expect(storeOptions.modules.Files.mutations.SET_HIDDEN_FILES_VISIBILITY).toHaveBeenCalled()
  })
  it('toggles the setting to show/hide file extensions', () => {
    const { wrapper, storeOptions } = getWrapper()
    ;(wrapper.findComponent<any>(selectors.fileExtensionsSwitch).vm as any).$emit('change', false)
    expect(storeOptions.modules.Files.mutations.SET_FILE_EXTENSIONS_VISIBILITY).toHaveBeenCalled()
  })
  it('initially does not show a viewmode switcher', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.find('[data-testid="viewmode-switch-buttons"]').exists()).toBeFalsy()
  })
  it('shows a viewmode switcher if more than one viewModes are passed', () => {
    const { wrapper } = getWrapper(
      {},
      {
        viewModes: [ViewModeConstants.condensedTable, ViewModeConstants.default]
      }
    )
    const viewModeSwitchButtons = wrapper.find('[data-testid="viewmode-switch-buttons"]')
    expect(viewModeSwitchButtons.html()).toMatchSnapshot()
  })
})

function getWrapper(
  { perPage = '100' } = {},
  props?: {
    viewModes: ViewMode[]
  }
) {
  jest.mocked(useRouteQueryPersisted).mockImplementation(() => ref(perPage))

  const storeOptions = { ...defaultStoreMockOptions }
  const store = createStore(storeOptions)
  const mocks = defaultComponentMocks()
  return {
    storeOptions,
    mocks,
    wrapper: mount(ViewOptions, {
      props: { ...props },
      global: {
        mocks,
        stubs: { OcButton: true, OcPageSize: false, OcSelect: true },
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
