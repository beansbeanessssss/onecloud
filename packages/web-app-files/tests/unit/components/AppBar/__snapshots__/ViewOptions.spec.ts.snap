// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewOptions component sets the correct initial files page limit 1`] = `
<div class="oc-flex oc-flex-middle">
  <!--v-if-->
  <oc-button-stub appearance="raw" arialabel="Display customization options of the files list" class="oc-my-s oc-p-xs" disabled="false" gapsize="medium" id="files-view-options-btn" justifycontent="center" size="medium" submit="button" type="button" variation="passive"></oc-button-stub>
  <div class="oc-drop oc-box-shadow-medium oc-rounded oc-width-auto" id="files-view-options-drop">
    <div class="oc-card oc-card-body oc-background-secondary oc-p-m">
      <ul class="oc-list oc-my-rm oc-mx-rm">
        <li class="files-view-options-list-item oc-mb-m">
          <span class="oc-switch">
            <span id="oc-switch-label-1">Show hidden files</span>
            <button aria-checked="false" aria-labelledby="oc-switch-label-1" class="oc-switch-btn" role="switch"></button>
          </span>
        </li>
        <li class="files-view-options-list-item oc-my-m">
          <span class="oc-switch">
            <span id="oc-switch-label-2">Show file extensions</span>
            <button aria-checked="false" aria-labelledby="oc-switch-label-2" class="oc-switch-btn" role="switch"></button>
          </span>
        </li>
        <li class="files-view-options-list-item oc-mt-m">
          <div class="oc-page-size files-pagination-size">
            <label class="oc-page-size-label" for="oc-page-size-3">Items per page</label>
            <oc-select-stub class="oc-page-size-select" clearable="false" disabled="false" filter="[Function]" id="oc-select-4" input-id="oc-page-size-3" loading="false" model-value="100" options="100,500" searchable="false"></oc-select-stub>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`ViewOptions component shows a viewmode switcher if more than one viewModes are passed 1`] = `
<div class="viewmode-switch-buttons oc-button-group oc-visible@s oc-mr-s">
  <oc-button-stub appearance="outline" arialabel="Switch to condensed table view" class="resource-table-condensed" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive"></oc-button-stub>
  <oc-button-stub appearance="outline" arialabel="Switch to default table view" class="resource-table" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive"></oc-button-stub>
</div>
`;
